client/lib/analytics/index.js

```diff
@@ -37,7 +38,61 @@ window.ga = window.ga || function() {
 };
 window.ga.l = +new Date();
 
-loadScript( '//stats.wp.com/w.js?56' ); // W_JS_VER
+function getUrlParameter( name ) {
+	name = name.replace( /[\[]/, '\\[' ).replace( /[\]]/, '\\]' );
+	const regex = new RegExp( '[\\?&]' + name + '=([^&#]*)' );
+	const results = regex.exec( location.search );
+	return results === null ? '' : decodeURIComponent( results[ 1 ].replace( /\+/g, ' ' ) );
+}
+
+function newAnonId() {
+	const randomBytesLength = 18; // 18 * 4/3 = 24 (base64 encoded chars)
+	let randomBytes = [];
+
+	if ( window.crypto && window.crypto.getRandomValues ) {
+		randomBytes = new Uint8Array( randomBytesLength );
+		window.crypto.getRandomValues( randomBytes );
+	} else {
+		for ( let i = 0; i < randomBytesLength; ++i ) {
+			randomBytes[ i ] = Math.floor( Math.random() * 256 );
+		}
+	}
+
+	return btoa( String.fromCharCode.apply( String, randomBytes ) );
+}
+
+function checkForBlockedTracks() {
+	if ( ! _loadTracksError ) {
+		return;
+	}
+
+	let _ut, _ui;
+
+	// detect stats blocking, and include identity from URL, user or cookie if possible
+	if ( _user && _user.get() ) {
+		_ut = 'wpcom:user_id';
+		_ui = _user.get().ID;
+	} else {
+		_ut = getUrlParameter( '_ut' ) || 'anon';
+		_ui = getUrlParameter( '_ui' );
+
+		if ( ! _ui ) {
+			const cookies = cookie.parse( document.cookie );
+			if ( cookies.tk_ai ) {
+				_ui = cookies.tk_ai;
+			} else {
+				_ui = newAnonId();
+				document.cookie = cookie.serialize( 'tk_ai', _ui );
+			}
+		}
+	}
+
+	loadScript( '/nostats.js?_ut=' + encodeURIComponent( _ut ) + '&_ui=' + encodeURIComponent( _ui ) );
+}
+
+loadScript( '//stats.wp.com/w.js?56', function( /* error */ ) {
+	_loadTracksError = true;
```

Comment:
Is it possible this could fail after `checkForBlockedTracks` is called? It might be better to actually just load `nostats.js` here?

Reply:
Originally I had that, but I needed to make sure that we checked for successful loading AFTER the Calypso middleware has booted which loads the user etc. so that we had the most accurate user identification.